Index: config/routes.rb
===================================================================
--- config/routes.rb	(revision 0)
+++ config/routes.rb	(working copy)
@@ -14,9 +14,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

 RedmineApp::Application.routes.draw do
-  root :to => 'welcome#index', :as => 'home'
+  root :to => 'issues#index', :as => 'home', :project_id => 'lifeguard'
+	
+  match 'welcome', :to => 'welcome#index'

   match 'login', :to => 'account#login', :as => 'signin', :via => [:get, :post]
   match 'logout', :to => 'account#logout', :as => 'signout', :via => [:get, :post]
